Bin/Snapshot.php

```diff
@@ -273,16 +300,54 @@ function () use ($tags, $newTag, $repositoryRoot, &$changelog) {
                             $title = $newTag;
                         }
 
-                        $changelog .=
-                            '# ' . $title . "\n\n" .
-                            Console\Processus::execute(
-                                'git --git-dir=' . $repositoryRoot . '/.git ' .
-                                    'log ' .
-                                        '--first-parent ' .
-                                        '--pretty="format:  * %s (%aN, %aI)" ' .
-                                        $fromStep . '...' . $toStep,
-                                false
-                            ) . "\n\n";
+                        $changelog .= '# ' . $title . "\n\n";
+
+                        $commits = Console\Processus::execute(
+                            'git --git-dir=' . $repositoryRoot . '/.git ' .
+                                'log ' .
+                                    '--first-parent ' .
+                                    '--pretty="format:  * %s (%aN, %aI)" ' .
+                                    $fromStep . '...' . $toStep,
+                            false
+                        );
+
+                        $collection = explode(PHP_EOL, $commits);
+                        $pretty = [];
+                        $otherCommits = [];
+                        $regex = '/(chore|feat|fix|test|doc|undef)\(([^)]+)\)(.*)/m';
+                        foreach ($collection as $line) {
```

Comment:
I would prefer evaluate `onlyStructuredCommits` before iterate, in order to avoid `preg_match` if it will not used.

```php
if (false === $onlyStructuredCommits) {
    $otherCommits = array_values($collection);
} else {
    foreach ($collection as $line) {
        //...
    }
}
```

Reply:
Hum, interesting remark...

For me the `onlyStructuredCommits` flag is here to decide if we need to keep the unstructured commits or not. But the structured commits must always been processed using the regex.

If we apply your suggested change, we have 2 modes to generate the changelog :

* as a flat commit list (without processing structured commits),
* as a structured list (but by dropping the non structured commits).

Is it the wanted behaviour ?

Reply:
I've pushed a little update to generate the regex from the `CHANGELOG_THEMES` constant. Your comment is still valid here.